File: Data Structures and Algorithms GFG/Chapter-6 Arrays/Topic-2 Operations on Arrays/Operations on Fixed Sized Arrays/Insertion/Program_1.c
/*C Program to Insert an Element at a Given
Index in a Fixed Sized Array(Heap Allocation)*/

#include<stdio.h>
#include<stdlib.h>
#include<limits.h>

/*Decalaring Global Variables*/
long long unsigned int capacity, size, position;
int *arr;

/*Function to Create a Fixed Sized
Array using Heap Allocation*/
short unsigned int create_a_fixed_sized_array(void)
{
    /*User Input for capacity*/
    printf("Enter the Capacity of Array:");
    scanf("%llu", &capacity);
    /*User is advised to enter a Non-Negative Integer that is
    greater than or equal to 0 and less than (2 raised to power 64)
    as capacity otherwise the result will not be as predicted*/

    /*Checking for bounds for capacity*/
    if(capacity>ULLONG_MAX)
    {
        printf("%llu is an Invalid Input for Capacity!\n", capacity);
        return 0;
    }

    /*User Input for size*/
    printf("Enter the Size of Array:");
    scanf("%llu", &size);
    /*User is advised to enter a Non-Negative Integer that is
    greater than or equal to 0 and less than (2 raised to power 64)
    as size otherwise the result will not be as predicted*/

    /*Checking for bounds for size*/
    if(size>ULLONG_MAX)
    {
        printf("%llu is an Invalid Input for size!\n", size);
        return 0;
    }

    /*Comparison of size and capacity*/
    if(size>capacity)
    {
        printf("Size can't be greater than Capacity!\n");
        return 0;
    }

    /*Allocating Memory to Array*/
    arr=(int *)malloc(capacity*sizeof(int));

    /*Checking for Memory Allocation*/
    if(arr==NULL)
    {
        printf("Memory Allocation Failed!\n");
        return 0;
    }

    /*Taking the Elements of Array as User Input*/
    printf("Enter %llu Elements for Array:\n", size);
    for(long long unsigned int i=0; i<size; i++)
    {
        printf("Enter Element %llu of Array:", i+1);
        scanf("%d", &arr[i]);
    }
    return 1;
}

/*Function to Display Elements of Array: Display Function*/
void display(void)
{
    /*Displaying the Elements of Array: Traversing an Array*/
    printf("The Elements of Array are:\n");
    for(long long unsigned int i=0; i<size; i++)
    printf("Element %llu of Array is:%d\n", i+1, i[arr]);
}

/*Function to Insert an Element at a Given Position in an Array*/
short unsigned int insert(int element)
{
    if(size==capacity)
    {
        printf("Can't Insert %d in Array at Position %llu because of Overflow!", element, position);
        return 0;
    }
    else if(position==0 || position>(size+1))
    {
        printf("%llu is an Invalid Input for Position!", position);
        return 0;
    }
    else
    {
        long long unsigned int index=position-1;
        long long unsigned int i;
        for(i=size-1; i>=index; i--)
        {
            arr[i+1]=arr[i];
            if(i==0)
            {
                arr[i+1]=arr[i];
                break;
            }
        }
        if(i==0)
        arr[i]=element;
        else
        arr[i+1]=element;
        size++;
        return 1;
    }
}

/*Driver Function(Code)*/
int main()
{
    if(!create_a_fixed_sized_array())
    printf("Array Creation Failed!");
    else
    {
        printf("Array before Insertion Operation:\n");
        display();
        printf("And Size of Array before Insertion Operation:%llu\n", size);
        int element;
        printf("Enter the Element you want to Insert in Array:");
        scanf("%d", &element);
        printf("Enter the Position at which you want to Insert %d:", element);
        scanf("%llu", &position);
        /*User is advised to enter a Non-Negative Integer that is
        greater than or equal to 0 and less than (2 raised to power 64)
        as position otherwise the result will not be as predicted*/

        if(!insert(element))
        printf("Insertion Failed!");
        else
        {
            printf("Array after Insertion Operation:\n");
            display();
            printf("And Size of Array after Insertion Operation:%llu\n", size);
        }
    }
    return 0;
}

File: Data Structures and Algorithms GFG/Chapter-6 Arrays/Topic-2 Operations on Arrays/Operations on Fixed Sized Arrays/Insertion/Program_1.java
/*Java Program to Insert an Element at a Given
Index in a Fixed Sized Array(Heap Allocation)*/

import java.util.*;

/*Creating an outer class*/
class Program_1{

    /*Creating an inner class*/
    class Implement_Fixed_Sized_Array{

        /*Data Members of class: int and int []*/
        int capacity, size, position;
        int []arr;
        int element;

        /*Defining readInput Function*/
        void readInput(Scanner sc)
        {
            System.out.println("Enter the Capacity of Array:");
            capacity=sc.nextInt();
            System.out.println("Enter the Size of Array:");
            size=sc.nextInt();
        }

        /*Member Function of class:
        to Create a Fixed Sized
        Array using Heap Allocation*/
        boolean create_a_fixed_sized_array(Scanner sc)
        {
            readInput(sc);
            /*User is advised to enter a Non-Negative Integer that is
            greater than or equal to 0 and less than (2 raised to power 63)
            as capacity and size otherwise the result will not be as predicted*/

            /*Checking for bounds for capacity*/
            if(capacity>Integer.MAX_VALUE || capacity<0)
            {
                System.out.println(capacity+" is an Invalid Input for Capacity!");
                return false;
            }

            /*Checking for bounds for size*/
            if(size>Integer.MAX_VALUE || size<0)
            {
                System.out.println(size+" is an Invalid Input for Size!");
                return false;
            }

            /*Comparison of size and capacity*/
            if(size>capacity)
            {
                System.out.println("Size can't be greater than Capacity!");
                return false;
            }

            /*Allocating Memory to Array*/
            try{
                arr=new int[capacity];
            }
            catch(OutOfMemoryError e){
                System.out.println("Memory Allocation Failed!");
            }

            /*Taking the Elements of Array as User Input*/
            System.out.println("Enter "+size+" Elements for Array:");
            for(int i=0; i<size; i++)
            {
                System.out.println("Enter Element "+(i+1)+" of Array:");
                arr[i]=sc.nextInt();
            }
            return true;
        }

        /*Member Function of class:
        to Display Elements of Array:
        Display Function*/
        void display()
        {
            /*Displaying the Elements of Array: Traversing an Array*/
            System.out.println("The Elements of Array are:");
            for(int i=0; i<size; i++)
            System.out.println("Element "+(i+1)+" of Array is:"+arr[i]);
        }

        /*Member Function of class:
        to Insert an Element at a Given 
        Position in an Array*/
        boolean insert(int element)
        {
            if(size==capacity)
            {
                System.out.println("Can't Insert "+element+" in Array at Position "+position+" because of Overflow!");
                return false;
            }
            else if(position<=0 || position>(size+1))
            {
                System.out.println(position+" is an Invalid Input for Position!");
                return false;
            }
            else
            {
                int index=position-1;
                int i;
                for(i=size-1; i>=index; i--)
                arr[i+1]=arr[i];    
                arr[i+1]=element;
                size++;
                return true;
            }
        }
    }

    /*Driver Function(Code)*/
    public static void main(String[] args)
    {
        Scanner sc=new Scanner(System.in);
        Program_1 program_1=new Program_1();
        Program_1.Implement_Fixed_Sized_Array IFSA=program_1.new Implement_Fixed_Sized_Array();
        if(!IFSA.create_a_fixed_sized_array(sc))
        System.out.println("Array Creation Failed!");
        else
        {
            System.out.println("Array before Insertion Operation:");
            IFSA.display();
            System.out.println("And Size of Array before Insertion Operation:"+IFSA.size);
            System.out.println("Enter the Element you want to Insert in Array:");
            IFSA.element=sc.nextInt();
            System.out.println("Enter the Position at which you want to Insert "+IFSA.element+":");
            IFSA.position=sc.nextInt();
            /*User is advised to enter a Non-Negative Integer that is
            greater than or equal to 0 and less than (2 raised to power 63)
            as position otherwise the result will not be as predicted*/

            if(IFSA.insert(IFSA.element)==false)
            System.out.println("Insertion Failed!");
            else
            {
                System.out.println("Array after Insertion Operation:");
                IFSA.display();
                System.out.println("And Size of Array after Insertion Operation:"+IFSA.size);
            }
        }
        sc.close();
    }
}
